src/core/credentials/transport/tls/grpc_tls_certificate_provider.cc

```diff
@@ -281,28 +318,54 @@ absl::Status FileWatcherCertificateProvider::ValidateCredentials() const {
       return status;
     }
   }
-  return absl::OkStatus();
+  // Validation of the SPIFFE Bundle happens when it is loaded from JSON, but
+  // the ctor of FileWatcherCertificateProvider doesn't allow for that error to
+  // bubble up. Do that here. If there is no bad status, the default of Ok will
+  // be passed up
+  return spiffe_load_status_;
 }
 
 void FileWatcherCertificateProvider::ForceUpdate() {
-  std::optional<std::string> root_certificate;
+  std::shared_ptr<RootCertInfo> root_cert_info;
   std::optional<PemKeyCertPairList> pem_key_cert_pairs;
-  if (!root_cert_path_.empty()) {
-    root_certificate = ReadRootCertificatesFromFile(root_cert_path_);
+  absl::Status spiffe_load_status;
+  if (!spiffe_bundle_map_path_.empty()) {
+    auto map = SpiffeBundleMap::FromFile(spiffe_bundle_map_path_);
+    if (map.ok()) {
+      root_cert_info = std::make_shared<RootCertInfo>(std::move(*map));
+      spiffe_load_status = absl::OkStatus();
+    } else {
+      spiffe_load_status = map.status();
+    }
+  } else if (!root_cert_path_.empty()) {
+    std::optional<std::string> root_certificate =
+        ReadRootCertificatesFromFile(root_cert_path_);
+    if (root_certificate.has_value()) {
+      root_cert_info =
+          std::make_shared<RootCertInfo>(std::move(*root_certificate));
+    }
   }
   if (!private_key_path_.empty()) {
     pem_key_cert_pairs = ReadIdentityKeyCertPairFromFiles(
         private_key_path_, identity_certificate_path_);
   }
   MutexLock lock(&mu_);
-  const bool root_cert_changed =
-      (!root_certificate.has_value() && !root_certificate_.empty()) ||
-      (root_certificate.has_value() && root_certificate_ != *root_certificate);
-  if (root_cert_changed) {
-    if (root_certificate.has_value()) {
-      root_certificate_ = std::move(*root_certificate);
+  spiffe_load_status_ = spiffe_load_status;
+  // If update has no value, but the existing root has a value, then the update
+  // is a delete
+  const bool is_root_update_a_delete =
+      root_cert_info == nullptr && root_cert_info_ != nullptr;
+  // If the update has a value, see if the existing value is nullptr or has a
+  // different value than the update.
+  const bool did_root_change_value =
+      root_cert_info != nullptr &&
+      (root_cert_info_ == nullptr || *root_cert_info != *root_cert_info_);
+  const bool root_changed = is_root_update_a_delete || did_root_change_value;
+  if (root_changed) {
+    if (root_cert_info != nullptr) {
```

Comment:
I think this can just be `root_cert_info_ = std::move(root_cert_info)` in either case.  If it's null, then you can still just use the value.